flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+
+		this.totalSize = 0;
+		this.requestCount = 0;
+
+		this.logicallyRemovedNodes = new HashSet<>();
+		this.pruningValueNodes = ConcurrentHashMap.newKeySet();
+
+		this.closed = new AtomicBoolean(false);
+		this.resourceGuard = new ResourceGuard();
+	}
+
+	@Override
+	public int size() {
+		return totalSize - logicallyRemovedNodes.size();
+	}
+
+	/**
+	 * Returns total size of this map, including logically removed state.
+	 */
+	public int totalSize() {
+		return totalSize;
+	}
+
+	public int getRequestCount() {
+		return requestCount;
+	}
+
+	@Override
+	public S get(K key, N namespace) {
+		updateStat();
+
+		long node = getNodeInternal(key, namespace);
+
+		if (node == NIL_NODE) {
+			return null;
+		}
+
+		return helpGetNodeState(node);
+	}
+
+	@Override
+	public boolean containsKey(K key, N namespace) {
+		updateStat();
+
+		long node = getNodeInternal(key, namespace);
+
+		return node != NIL_NODE;
+	}
+
+	@Override
+	public void put(K key, N namespace, S state) {
+		updateStat();
+		ByteBuffer keyByteBuffer = getKeyByteBuffer(key, namespace);
+		int keyLen = keyByteBuffer.limit();
+		byte[] value = skipListValueSerializer.serialize(state);
+
+		putNode(keyByteBuffer, 0, keyLen, value, false);
+	}
+
+	@Override
+	public S putAndGetOld(K key, N namespace, S state) {
+		updateStat();
+		ByteBuffer keyByteBuffer = getKeyByteBuffer(key, namespace);
+		int keyLen = keyByteBuffer.limit();
+		byte[] value = skipListValueSerializer.serialize(state);
+
+		return putNode(keyByteBuffer, 0, keyLen, value, true);
+	}
+
+	@Override
+	public void remove(K key, N namespace) {
+		updateStat();
+		ByteBuffer keyByteBuffer = getKeyByteBuffer(key, namespace);
+		int keyLen = keyByteBuffer.limit();
+
+		removeNode(keyByteBuffer, 0, keyLen, false);
+	}
+
+	@Override
+	public S removeAndGetOld(K key, N namespace) {
+		updateStat();
+		ByteBuffer keyByteBuffer = getKeyByteBuffer(key, namespace);
+		int keyLen = keyByteBuffer.limit();
+
+		return removeNode(keyByteBuffer, 0, keyLen, true);
+	}
+
+	@Override
+	public <T> void transform(
+		K key,
+		N namespace,
+		T value,
+		StateTransformationFunction<S, T> transformation) throws Exception {
+		updateStat();
+		ByteBuffer keyByteBuffer = getKeyByteBuffer(key, namespace);
+		int keyLen = keyByteBuffer.limit();
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+		S oldState = node == NIL_NODE ? null : helpGetNodeState(node);
+		S newState = transformation.apply(oldState, value);
+		byte[] stateBytes = skipListValueSerializer.serialize(newState);
+		putNode(keyByteBuffer, 0, keyLen, stateBytes, false);
+	}
+
+	// Detail implementation methods ---------------------------------------------------------------
+
+	/**
+	 * Find the node containing the given key.
+	 *
+	 * @param keyByteBuffer byte buffer storing the key.
+	 * @param keyOffset     offset of the key.
+	 * @param keyLen        length of the key.
+	 * @return id of the node. NIL_NODE will be returned if key does no exist.
+	 */
+	@VisibleForTesting
```

Comment:
Because we want to test the non-zero `keyOffset` case in `CopyOnWriteSkipListStateMapTest.testPutAndGetNode`

Reply:
Can't we remove `keyOffset` and alway set it to `0`? It seems that we include functionality which is only used in tests because there is no internal caller which provides a non-zero `keyOffset`. I think in these cases it is usually better to remove dead functionality.

Reply:
Only allow the offset to be 0 sounds like a too hard limitation to me and may block us from future optimization, like taking use of a partially available byte buffer.